masonry/src/widgets/indexed_stack.rs

```diff
@@ -0,0 +1,382 @@
+    }
+
+    /// Builder-style method to add a child widget already wrapped in a [`WidgetPod`].
+    pub fn with_child_pod(mut self, widget: WidgetPod<dyn Widget>) -> Self {
+        self.children.push(widget);
+        self
+    }
+
+    /// Builder-style method to set the active child
+    ///
+    /// Index must be a valid index into the stack's children
+    pub fn with_active_child(mut self, idx: usize) -> Self {
+        assert!(self.children.is_empty() || idx < self.children.len());
+        self.active_child = idx;
+        self
+    }
+
+    /// Returns the number of children in this stack
+    pub fn len(&self) -> usize {
+        self.children.len()
+    }
+
+    /// Returns true if there are no children in this stack
+    pub fn is_empty(&self) -> bool {
+        self.children.is_empty()
+    }
+}
+
+// --- MARK: IMPL WIDGETMUT ---
+impl IndexedStack {
+    /// Add a child widget.
+    ///
+    /// See also [`with_child`](IndexedStack::with_child).
+    pub fn add_child(this: &mut WidgetMut<'_, Self>, child: impl Widget) {
+        let child_pod: WidgetPod<dyn Widget> = WidgetPod::new(child).erased();
+        Self::add_child_pod(this, child_pod);
+    }
+
+    /// Add a child widget with a pre-assigned id.
+    ///
+    /// See also [`with_child_id`](IndexedStack::with_child_id).
+    pub fn add_child_id(this: &mut WidgetMut<'_, Self>, child: impl Widget, id: WidgetId) {
+        let child_pod: WidgetPod<dyn Widget> = WidgetPod::new_with_id(child, id).erased();
+        Self::add_child_pod(this, child_pod);
+    }
+
+    /// Add a child widget already wrapped in a [`WidgetPod`].
+    ///
+    /// See also [`with_child_pod`](IndexedStack::with_child_pod).
+    pub fn add_child_pod(this: &mut WidgetMut<'_, Self>, widget: WidgetPod<dyn Widget>) {
+        this.widget.children.push(widget);
+        this.ctx.children_changed();
+        this.ctx.request_layout();
+    }
+
+    /// Insert a child widget at the given index.
+    ///
+    /// This lets you control the order of the children stored by the indexed stack.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the index is larger than the number of children.
+    pub fn insert_stack_child_at(this: &mut WidgetMut<'_, Self>, idx: usize, child: impl Widget) {
+        Self::insert_stack_child_pod(this, idx, WidgetPod::new(child).erased());
+    }
+
+    /// Insert a child widget already wrapped in a [`WidgetPod`] at the given index.
+    ///
+    /// This lets you control the order of the children stored by the indexed stack.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the index is larger than the number of children.
+    pub fn insert_stack_child_pod(
+        this: &mut WidgetMut<'_, Self>,
+        idx: usize,
+        child: WidgetPod<dyn Widget>,
+    ) {
+        this.widget.children.insert(idx, child);
+        if this.widget.active_child >= idx {
+            // adjust index to keep the same widget active
+            this.widget.active_child += 1;
```

Comment:
I sent my review a bit early; I think perhaps the most reasonable way to make this consistent is to never change the active index in `insert_stack_child` or `remove_stack_child`.

This will be hard to document, but will avoid hard to reason about cases - the index will then always be the value you have set it to.